File: src/shared/api/llmConfig.ts
import type { LLMProvider } from "../../store/settings";
import type { LlmConfig } from "./apiClient";
import type { EmbeddingProvider } from "../../store/settings";

export const LOCAL_LLM_BASE_URL = "http://localhost:1234/v1";

export interface LlmConfigOverrides {
  maxTokens?: number;
  temperature?: number;
  baseUrl?: string;
  model?: string;
}

const providerLabels: Record<LLMProvider, LlmConfig["provider"]> = {
  local: "local",
  openai: "openai",
  gemini: "gemini",
  ollama: "ollama",
  llama_cpp: "llama_cpp",
  dll: "dll",
};

export function createLlmConfig(
  settings: {
    provider: LLMProvider;
    model: string;
    apiKey: string;
    baseUrl: string;
  },
  overrides: LlmConfigOverrides = {}
): LlmConfig {
  const isEmbeddedLocal = settings.provider === "local";
  const isKeyless =
    settings.provider === "local" ||
    settings.provider === "ollama" ||
    settings.provider === "llama_cpp";
  return {
    provider: providerLabels[settings.provider],
    base_url: isEmbeddedLocal
      ? LOCAL_LLM_BASE_URL
      : overrides.baseUrl ?? settings.baseUrl,
    model: overrides.model ?? settings.model,
    api_key: isKeyless ? null : settings.apiKey || null,
    max_tokens: overrides.maxTokens ?? 1024,
    temperature: overrides.temperature ?? 0.7,
  };
}

export function createEmbeddingConfig(settings: {
  provider: EmbeddingProvider;
  model: string;
}): LlmConfig {
  return {
    provider: providerLabels[settings.provider as LLMProvider],
    base_url: "",
    model: settings.model,
    api_key: null,
    max_tokens: 1024,
    temperature: 0.7,
  };
}

File: src-tauri/src/domain/llm_config.rs
use serde::{Deserialize, Serialize};
use std::fmt;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum LLMProvider {
    Local,
    OpenAI,
    #[serde(alias = "google", alias = "Google")]
    Gemini,
    #[serde(alias = "ollama")]
    Ollama,
    #[serde(alias = "llama_cpp")]
    LlamaCpp,
    #[serde(alias = "openrouter")]
    OpenRouter,
    #[serde(alias = "dll", rename = "dll")]
    Dll,
}

impl fmt::Display for LLMProvider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LLMProvider::Local => write!(f, "local"),
            LLMProvider::OpenAI => write!(f, "openai"),
            LLMProvider::Gemini => write!(f, "gemini"),
            LLMProvider::Ollama => write!(f, "ollama"),
            LLMProvider::LlamaCpp => write!(f, "llama_cpp"),
            LLMProvider::OpenRouter => write!(f, "openrouter"),
            LLMProvider::Dll => write!(f, "dll"),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct LLMConfig {
    pub provider: LLMProvider,
    #[serde(alias = "baseUrl")]
    pub base_url: String,
    pub model: String,
    #[serde(alias = "apiKey")]
    pub api_key: Option<String>,
    #[serde(alias = "maxTokens")]
    pub max_tokens: Option<u32>,
    pub temperature: Option<f32>,
}

/// Chat message with role and content
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

impl Default for LLMConfig {
    fn default() -> Self {
        Self {
            provider: LLMProvider::Local,
            base_url: "http://localhost:11434".to_string(),
            model: "nomic-embed-text".to_string(),
            api_key: None,
            max_tokens: Some(1024),
            temperature: Some(0.7),
        }
    }
}
